Mathlib/Algebra/Order/Archimedean/Class.lean

```diff
@@ -451,4 +461,183 @@ theorem liftOrderHom_mk (f : M → α) (h : ∀ a b, mk a ≤ mk b → f a ≤ f
 
 end LiftHom
 
+/-- Given a non-empty `UpperSet` of `MulArchimedeanClass`,
+all group elements belonging to these classes form a subgroup. -/
+@[to_additive "Given a non-empty `UpperSet` of `ArchimedeanClass`,
+all group elements belonging to these classes form a subgroup."]
+def subgroup' {s : UpperSet (MulArchimedeanClass M)} (hs : s ≠ ⊤) : Subgroup M where
+  carrier := mk ⁻¹' s
+  mul_mem' {a b} ha hb := by
+    rw [Set.mem_preimage] at ha hb ⊢
+    obtain h | h := min_le_iff.mp (min_le_mk_mul a b)
+    · exact s.upper h ha
+    · exact s.upper h hb
+  one_mem' := by
+    rw [Set.mem_preimage]
+    obtain ⟨u, hu⟩ := UpperSet.coe_nonempty.mpr hs
+    simpa using s.upper (by simp) hu
+  inv_mem' {a} h := by simpa using h
+
+/-- Extend `MulArchimedeanClass.subgroup'` to any `UpperSet`
+with a junk value `⊥` assigned to the empty set. -/
+@[to_additive "Extend `ArchimedeanClass.subgroup'` to any `UpperSet`
+with a junk value `⊥` assigned to the empty set."]
+noncomputable
+def subgroup (s : UpperSet (MulArchimedeanClass M)) : Subgroup M :=
+  open Classical in
+  if hs : s = ⊤ then
+    ⊥
+  else
+    subgroup' hs
```

Comment:
Is there any value in having `subgroup'` given that it's always equal to `subgroup` when defined?